django/db/backends/mysql/operations.py

```diff
@@ -175,21 +175,54 @@ def random_function_sql(self):
 
     def sql_flush(self, style, tables, sequences, allow_cascade=False):
         # NB: The generated SQL below is specific to MySQL
-        # 'TRUNCATE x;', 'TRUNCATE y;', 'TRUNCATE z;'... style SQL statements
+        # 'TRUNCATE x;', 'TRUNCATE y;', 'DELETE FROM z;'... style SQL statements
         # to clear all tables of all data
+        # 'DELETE FROM' is used for small tables, 'TRUNCATE' is used for large tables
+
         if tables:
+            schema_name = self.connection.settings_dict['NAME']
+
+            with self.connection.cursor() as cursor:
+                cursor.execute("""
+                SELECT table_name, table_rows
+                FROM information_schema.tables
+                WHERE table_schema = %s AND table_name IN %s
+                """, (schema_name, tables))
+                rows = cursor.fetchall()
+
             sql = ['SET FOREIGN_KEY_CHECKS = 0;']
-            for table in tables:
-                sql.append('%s %s;' % (
-                    style.SQL_KEYWORD('TRUNCATE'),
-                    style.SQL_FIELD(self.quote_name(table)),
-                ))
+
+            for table_name, table_rows in rows:
+                if table_rows <= 1000:
+                    # 1000 is a heuristic threshold to determine
+                    # whether using "DELETE FROM" or "TRUNCATE".
+                    sql.append('%s %s %s;' % (
+                        style.SQL_KEYWORD('DELETE'),
+                        style.SQL_KEYWORD('FROM'),
+                        style.SQL_FIELD(self.quote_name(table_name)),
+                    ))
+                else:
+                    sql.append('%s %s;' % (
+                        style.SQL_KEYWORD('TRUNCATE'),
+                        style.SQL_FIELD(self.quote_name(table_name)),
+                    ))
+
             sql.append('SET FOREIGN_KEY_CHECKS = 1;')
             sql.extend(self.sequence_reset_by_name_sql(style, sequences))
             return sql
         else:
             return []
 
+    def sequence_reset_by_name_sql(self, style, sequences):
+        return [
+            '%s %s %s %s = 1;' % (
+                style.SQL_KEYWORD('ALTER'),
+                style.SQL_KEYWORD('TABLE'),
+                style.SQL_FIELD(self.quote_name(sequence_info['table'])),
+                style.SQL_FIELD('AUTO_INCREMENT'),
+            ) for sequence_info in sequences
+        ]
```

Comment:
We dropped `sequence_reset_by_name_sql()` intentionally in d28396f5268f1974ef1e84d13bcf1ac107005ced. It's redundant after `TRUNCATE`, we should call it only in `DELETE FROM` branch.

Reply:
Hi. Thank you for your review. I'm thinking about how should I remove redundant sequence resets after `TRUNCATE`.

One solution is to select `auto_increment` field from `information_schema.tables` at `sequence_reset_by_name_sql()`, or revert df2177682ad1. Do you have any good ideas?

Reply:
The former solution is like this:

```python
    def sequence_reset_by_name_sql(self, style, sequences):
        tables = [sequence_info['table'] for sequence_info in sequences]
        with self.connection.cursor() as cursor:
            cursor.execute("""
            SELECT table_name
            FROM information_schema.tables
            WHERE table_schema = DATABASE()
              AND table_name IN %s
              AND auto_increment IS NOT NULL
              AND auto_increment > 1
            """, (tables,))
            rows = cursor.fetchall()

        return [
            '%s %s %s %s = 1;' % (
                style.SQL_KEYWORD('ALTER'),
                style.SQL_KEYWORD('TABLE'),
                style.SQL_FIELD(self.quote_name(row[0])),
                style.SQL_FIELD('AUTO_INCREMENT'),
            ) for row in rows
        ]
```